keps/sig-node/1287-in-place-update-pod-resources/README.md

```diff
@@ -99,48 +101,89 @@ Other identified non-goals are:
+larger of `Spec.Containers[i].Resources` and
+`Status.ContainerStatuses[i].ResourcesAllocated` when considering available
+space on a node.
 
-A new admission controller named 'PodResourceAllocation' is introduced in order
-to limit access to ResourcesAllocated field such that only Kubelet can update
-this field.
+#### Subresource
 
-Additionally, Kubelet is authorized to update PodSpec, and NodeRestriction
-admission plugin is extended to limit Kubelet's update access only to Pod's
-ResourcesAllocated field for CPU and memory resources.
+For alpha, resource changes will be made by updating the pod spec.  For beta
+(or maybe a followup in alpha), a new subresource, /resize, will be defined.
+This subresource could eventually apply to other resources that carry
+PodTemplates, such as Deployments, ReplicaSets, Jobs, and StatefulSets.  This
+will allow users to grant RBAC access to controllers like VPA without allowing
+full write access to pod specs.
+
+The exact API here is TBD.
 
 #### Container Resize Policy
 
 To provide fine-grained user control, PodSpec.Containers is extended with
-ResizePolicy - a list of named subobjects (new object) that supports 'cpu'
-and 'memory' as names. It supports the following policy values:
-* NoRestart - the default value; resize Container without restarting it,
-* RestartContainer - restart the Container in-place to apply new resource
-  values. (e.g. Java process needs to change its Xmx flag)
-
-By using ResizePolicy, user can mark Containers as safe (or unsafe) for
-in-place resource update. Kubelet uses it to determine the required action.
+ResizePolicy - a list of named subobjects (new object) that supports 'cpu' and
+'memory' as names. It supports the following policy values:
+* RestartNotRequired - default value; try to resize the Container without
+  restarting it, if possible.
+* Restart - the container requires a restart to apply new resource values.
+  (e.g.  Java process needs to change its Xmx flag) By using ResizePolicy, user
+  can mark Containers as safe (or unsafe) for in-place resource update. Kubelet
+  uses it to determine the required action.
+
+Note: `RestartNotRequired` does not guarantee that container won't be
+restarted. The runtime may choose to stop the container if it is unable to
+apply the new resources without doing so.
 
 Setting the flag to separately control CPU & memory is due to an observation
 that usually CPU can be added/removed without much problem whereas changes to
 available memory are more probable to require restarts.
 
-If more than one resource type with different policies are updated, then
-RestartContainer policy takes precedence over NoRestart policy.
+If more than one resource type with different policies are updated at the same
+time, then any `Restart` policy takes precedence over `RestartNotRequired` policies.
+
+If a container's RestartPolicy is `Never`, the ResizePolicy must be set to
```

Comment:
nit: RestartPolicy is at PodSpec level, applicable to all containers in the Pod.

Reply:
fixed